### PR TITLE
Cleanup C++ compiler macros usage

### PR DESCRIPTION
**Pull request by externl**
This PR is a backport of https://github.com/zeroc-ice/ice/commit/158c19368c33ca12def844149614f47a128d60db

### CODE REVIEW

File: cpp/include/IceUtil/Config.h [hunk 1]
```diff
@@ -109,7 +109,7 @@
 //
 #if (ICE_CPLUSPLUS >= 201103) || \
     ((defined(__GNUC__) && defined(__GXX_EXPERIMENTAL_CXX0X__) && ((__GNUC__* 100) + __GNUC_MINOR__) >= 405)) || \
```
Comment (pepone) [inline cpp/include/IceUtil/Config.h:111]: Not clear that we need to keep this branch with `__GXX_EXPERIMENTAL_CXX0X__`

File: cpp/src/IceUtil/UtilException.cpp [hunk 2]
```diff
@@ -43,7 +43,7 @@
 #       endif
 #   endif
 
-#   if !defined(_AIX) && !defined(__sun) && !defined(__FreeBSD__) && !defined(__MINGW32__) && !defined(ICE_STATIC_LIBS)
+#   if !defined(_AIX) && !defined(__sun) && !defined(__FreeBSD__) && !defined(ICE_STATIC_LIBS)
```
Comment (pepone) [inline cpp/src/IceUtil/UtilException.cpp:46]: We can also remove `!defined(__sun) ` as we removed other Sun related macros

File: cpp/src/IceUtil/UtilException.cpp [hunk 3]
```diff
@@ -54,15 +54,11 @@
...
-#   if defined(_MSC_VER) && (_MSC_VER >= 1700)
-#       define DBGHELP_TRANSLATE_TCHAR
-#       include <IceUtil/StringConverter.h>
-#       if _MSC_VER >= 1900
-#           pragma warning(disable:4091) // VS 2015 RC issues this warning for code in DbgHelp.h
-#       endif
-#   endif
+#   define DBGHELP_TRANSLATE_TCHAR
+#   include <IceUtil/StringConverter.h>
+#   pragma warning(disable:4091) // VS 2015 RC issues this warning for code in DbgHelp.h
```
Comment (pepone) [inline cpp/src/IceUtil/UtilException.cpp:61]: Can you add a TODO for this one, not clear it is required with Visual Studio 2022

**Review by pepone — APPROVED** [review]
Looks good!

File: cpp/include/IceUtil/Config.h (hunk 1 above)
Comment (externl) [reply cpp/include/IceUtil/Config.h:111]: We'll want to clean this up anyway soon since all our compilers will be c++11 at least.

Comment (externl) [reply cpp/include/IceUtil/Config.h:111]: I'll leave it like it is for now.

**Comment by externl** [comment]
I also removed Atomic.h

**Pull request merged**